lld/wasm/InputFiles.cpp

```diff
@@ -408,6 +408,12 @@ ObjFile::ObjFile(MemoryBufferRef m, StringRef archiveName, bool lazy)
   this->lazy = lazy;
   this->archiveName = std::string(archiveName);
 
+  // Currently we only do this check for regular object file, and not for shared
+  // object files.  This is because architecture detection for shared objects is
+  // currently based on a heuristic, which is fallable:
+  // https://github.com/llvm/llvm-project/issues/98778
+  checkArch(wasmObj->getArch());
```

Comment:
Does this problem (that we cannot determine arch if there is no memory) only happen with shared objects? If so, why? If non-shared objects don't have a memory, how do we determine the arch?

Reply:
Yes, non-shared objects, produced by llvm always import the memory.   We just get lucky that is the case.

In fact even shared object do, just not once they have been though wasm-opt, which will eliminate the memory completely if its unused.